components/global/core/io.cellery.observability.telemetry.deduplicator/src/main/java/io/cellery/observability/telemetry/deduplicator/DeduplicationStreamProcessor.java

```diff
@@ -0,0 +1,391 @@
+    private volatile long lastTimestamp = 0;
+
+    private ExpressionExecutor traceIdExecutor;
+    private ExpressionExecutor spanIdExecutor;
+    private ExpressionExecutor parentSpanIdExecutor;
+    private ExpressionExecutor sourceNamespaceExecutor;
+    private ExpressionExecutor sourceInstanceExecutor;
+    private ExpressionExecutor sourceComponentExecutor;
+    private ExpressionExecutor destinationNamespaceExecutor;
+    private ExpressionExecutor destinationInstanceExecutor;
+    private ExpressionExecutor destinationComponentExecutor;
+    private ExpressionExecutor requestSizeExecutor;
+    private ExpressionExecutor responseDurationExecutor;
+    private ExpressionExecutor responseSizeExecutor;
+
+    @Override
+    protected void process(ComplexEventChunk<StreamEvent> streamEventChunk, Processor nextProcessor,
+            StreamEventCloner streamEventCloner, ComplexEventPopulater complexEventPopulater) {
+        ComplexEventChunk<StreamEvent> nextProcessorEventChunk = new ComplexEventChunk<>(false);
+        synchronized (this) {
+            while (streamEventChunk.hasNext()) {
+                StreamEvent streamEvent = streamEventChunk.next();
+                long currentTime = siddhiAppContext.getTimestampGenerator().currentTime();
+                StreamEvent clonedEvent = null;
+                if (streamEvent.getType() == StreamEvent.Type.CURRENT) {
+                    clonedEvent = streamEventCloner.copyStreamEvent(streamEvent);
+                    clonedEvent.setType(StreamEvent.Type.EXPIRED);
+                }
+                expiredEventChunk.reset();
+                boolean isDuplicationEventFound = false;
+                while (expiredEventChunk.hasNext()) {
+                    StreamEvent expiredEvent = expiredEventChunk.next();
+                    long timeDiff = expiredEvent.getTimestamp() + windowTimeMilliSeconds - currentTime;
+                    if (streamEvent.getType() == StreamEvent.Type.CURRENT) {
+                        if (checkForDuplications(expiredEvent, clonedEvent)) {
+                            isDuplicationEventFound = true;
+                            this.expiredEventChunk.remove();
+                            nextProcessorEventChunk.add(mergeEvents(expiredEvent, clonedEvent));
+                        }
+                    } else if (timeDiff <= 0) {
+                        expiredEventChunk.remove();
+                        expiredEvent.setType(StreamEvent.Type.CURRENT);
+                        nextProcessorEventChunk.add(mergeEvents(null, expiredEvent));
+                        expiredEventChunk.reset();
+                    } else {
+                        break;
+                    }
+                }
+                if (!isDuplicationEventFound && streamEvent.getType() == StreamEvent.Type.CURRENT) {
+                    this.expiredEventChunk.add(clonedEvent);
+                    if (lastTimestamp < Objects.requireNonNull(clonedEvent).getTimestamp()) {
```

Comment:
Let's not use `Objects.requireNonNull()` as it will throw null pointer exception causing the stack to unwind.

Anyway clonedEvent cannot be null is the streamEvent is a current event.